### ISSUE TITLE
Make access_data values for accesses other than read uint64_t optional in access_log

### ISSUE THREAD

**Issue by diegonehab** [cartesi/machine-emulator#98]
## Context

At present, each `access` in an `access_log` returned by `machine::step_uarch()` contains two `access_data` fields `read` and `written`, as well as a `log2_size` specifying the power-of-two size of that access. Read accesses define only the `read` field. Write accesses define both `read` and `written` fields. 

In the past, these fields were `uint64_t`: The larger access types were added to support the dehashing device.

As it turns out, these larger accesses are also needed for the implementation of `machine::reset_uarch_state()` in Solidity. For example, the entire contents of uarch RAM must be restored back to its pristine condition. This could be represented as a single `write` access of the entire uarch RAM contents.

Interestingly, there was never a need for the `access_data` with the contents of what was actually written. In the case of the dehashing device, only the hash was needed (the data itself would be presented as proof that a block with that given hash existed). Likewise, in the case of `machine::reset_uarch_state()`, the hash of the uarch RAM must be hardcoded in Solidity anyways. 

In fact, even for those `access_data` that are read, only the `uint64_t` are ever needed. There doesn't seem to be a situation in which Solidity code must read large chunks of the state. And for write accesses, we only need the hash of what was previously there to validate the list of sibling hashes that will be used to stitch the hash of the newly written data into the Merkle tree. 

As currently implemented,  `access_log` makes unnecessary `access_data` mandatory. This is even more problematic because these `access_data` can be large (e.g., the uarch RAM could be megabytes of data). 

Nevertheless, having the option of recording these fields in the `access_log` for debugging purposes may be advantageous. 

## Possible solutions

The idea is to, by default, omit the `access_data` for the `written` field, an omit it for `read` fields where `log2_size` != 8. 

## Subtasks

- [ ] Change `access_log::type` to include a new option `verbose` that, when set, includes the unnecessary `access_data`;
- [ ] Change `machine::step_uarch()` to honor the option.

**Comment by GCdePaula** [comment]
Cool! Besides the needed changes, I like the insights on what could be omitted. But I'm a bit torn. I think there may be things mixed up. And I think the root of it is our typing/schema.

Since we're already changing some stuff, I want to give my take on how we could type things. I just want to be clear that these are just possible ideas; I'm not sure whether I'm missing something, or whether this is an improvement. 

Currently, it seems our logs always have the same fields, but depending on what was done (_i.e._ read word, read range, write) some fields come back nil and other's don't. Furthermore, `access_data` can have eight bytes or multiple megabytes of data, also depending on the kind (and even more curious, only the small sized `access_data` is useful; when the field is huge, its content is actually never read).

I'd argue for better typing. I see three distinct kinds of accesses: 
* reading a word (needs address, value of the word itself and proofs);
* reading a memory range (needs address, size, hash and proofs);
* writing (needs address, size, old hash, new hash and proofs);

I'd write types closer to this (using sum types; if not available use something closer to tagged unions):

```
struct MerkleProof {
    siblings: [bytes32]
}

struct MemoryRegion {
    address: u64,
    log2_size: u8,

    // There's a different way to specify a memory range instead of the pair `address` and `log2_size`.
    // Note that the pair is byte-sized; since we always make at least word-size access, the lower 3 bits of `address`
    // are always zero, and `log2_size` starts at 3. Also, the access has to be aligned to a power of two.
    // As such, one can specify an invalid pair `address` and `log2_size`.

    // We can use the pair `stride_count` and `log2_stride_size` (names TBD), which are always aligned (although then can
    // be out of bounds); `stride_count` means the number of times we "walked" a `stride_size`.
    // To convert them to `address`, use the formula `(stride_count * (2 ^ log2_stride_size)) << 3`.
    // To convert them to `log2_size`, use the formula `log2_stride_size + 3`.

    // Reiterating that either way is fine, this is just something we came up with and wanted to document.
}


enum Access {
    Write {
        region: MemoryRegion,
        old_value: bytes32,
        new_value: bytes32,
        proof: MerkleProof
    }

    ReadRange {
        region: MemoryRegion,
        value: bytes32,
        proof: MerkleProof
    }

    ReadWord {
        address: u64,
        value: u64,
        proof: MerkleProof
    }
}
```

Now, moving on to omitted fields. The Solidity code doesn't need the `new_value` field when data is written. But it also doesn't need `MemoryRegions` at all! And neither knowing the kind of access (_i.e._ which enum variant it is). If we strip all data that Solidity doesn't need, we'd end up with a buffer, filled with hashes and ocasional words, all concatenated; Solidity will know which is which.

So I think we should keep things information rich, and the node will strip away and encode things; with this new suggested typing schema, size is not an issue since there's no longer an `access_data` field that can vary from being word size to megabytes of data.

If you do want to give the possibility of a verbose log, I'd increment the enum variants with optional data.

Thoughts?

**Comment by diegonehab** [comment]
Setting aside how Rusty this is for a moment, I'll indulge you.

I don't see why we would duplicate the proof in each variant. After all, as far as Solidity is concerned, no matter the variant, the first thing you do is verify that the proof is valid. I.e., that the siblings and the target hash can be used to reconstruct the root hash. It seems desirable to factor that part of the code out of the treatment of each variant. You can't do that in an elegant way if the proof is inside the variants, rather than outside.

I am also not sure a range access is different enough from a word access to merit a different variant. The conversion from little-endian to big-endian should happen later, not in the access. Otherwise, when you compute the target hash for the value in a read access, to check that it matches what was provided in the proof, you'd have to convert back to little-endian. Feels unacceptable to me. Since the value will be little-endian, it is not really an u64, right? Especially in Solidity. It is simply an array of bytes that happens to have 8 bytes in it. The conversion happens only when you choose to interpret that as a number, later on. This makes the word access way more similar to the range access than what is implied by separate variants.

One difference between them is that, in the vast majority of cases, you will need the value for the word access, but not for the range access. But that is not very future proof. I can imagine some time in the future where we want larger read accesses that need the value to be there. A 32-byte access comes to mind as being relevant to our domain. Additionally, perhaps the easiest way to implement the dehashing device is for the pre-image to be sent along with the access itself. It's a bit more complicated than that (due to the nature of disputes involving the dehashing device), but I am not certain this is false.

I can't think of a situation where you need the value for a write access though, other than for debugging.

So I'd move the proof out and make it optional. I'd move the read value out, and also make it optional. The address and size would also be shared between all variants. So I'd move them out as well. The only thing that really is different is that write accesses have an optional written value. But is that really enough to warrant an empty read variant and a write variant with that written value that needs to be optional anyway? 

Not feeling it.

PS: There is an unfortunate duplication of information in that the proof also contains the `address` and `log2_size` of the node in question. If the proof was _not_ optional, we could remove the repeated `address` and `log2_size` either from the proof or from outside it. But we really need it to be optional or some of our tests would take a lot, lot longer or would have to include a bogus proof. Removing `address` and `log2_size` from the proof itself would make it not self-contained, which would also lead to uglier code elsewhere.

**Comment by GCdePaula** [comment]
> Setting aside how Rusty this is for a moment, I'll indulge you.

Handy syntax for algebraic data types! I felt using a different syntax could be a little pretentious heh. Thank your for the indulgence; hopefully it will clean me of my sins. I'm joking.

---
> I don't see why we would duplicate the proof in each variant. (...)

True! Hoist it away.

---
> I am also not sure a range access is different enough from a word access to merit a different variant. (...)

I disagree! I think they're fundamentally different. One is an operation on a Merkle tree, and the other is on an address space. It just so happens that our address space is eventually merkleized. A Merkle tree read will give us a hash, whereas reading an address space will give us memory -- in this specific case a word-sized chunk. This is concretely manifested on the type I wrote.

Also, we know what is the log2_size (which I removed from the variant since it's known), and consequently the size of the accessed data. It is true that typing it as a `u64` (little-endian 64-bit unsigned integer) is unwarranted. Who know what is the type of the value we just read. It could be a piece of a string, or a float, or integers of different endianness. It's a word, which fits in a `u64`. So I don't see this variant at all as an "an array of bytes that happens to have 8 bytes in it". I see it as a word.

Now I understand you may approach this differently. I'm probably seeing this difference between Merkle trees and address spaces because from Solidity I only see Merkle trees, and from C++ you only see memory. But I still think there's a difference there, even if it's just a special case of read. Since it is the most common kind of read, I'd create a special case just for words.

---

> I'd move the proof out and make it optional.

Sure.

> I'd move the read value out, and also make it optional.

In case of word reads, we know the value is a word. This is the most common read case. In case of writes, we don't care about that (just the hash, to rebuild the tree). In case of range reads, currently we just need the hash, but I agree it may change. I don't think this should be hoisted.

> The address and size would also be shared between all variants.

Word reads we know the size already. Like, it's a tiny redundancy, but I feel it's more elegant. Actually, I see the `MemoryRegion` type I described as specifying a node in a Merkle tree. And in the word read case, I see the address as an actual memory address.

So something like this:

```
enum Access {
    proof: Option<Siblings>,

    Write {
        region: MemoryRegion,
        old_merkle_hash: bytes32,
        old_value: Option<Bytes>,
        written_merkle_hash: bytes32,
        written_value: Option<Bytes>
    }

    ReadRange {
        region: MemoryRegion,
        merkle_hash: bytes32,
        value: Option<Bytes>
    }

    ReadWord {
        address: u64,
        value: bytes8
    }
}
```

---

Now, I took a look at the proof schema, and I saw that it is not just the siblings, but that it also contains the target hash and the address/size. I was thinking that it was just the siblings. With that in mind, I don't think what I wrote is ideal, it would be effectively replicating everything.

Like, the argument for what I described would just be around the `access_data` field, regarding its size and being optional. So I like what you described, I think it's better :)

So from the point of view of Solidity:
* If it's a word read (most common), we'll look at the `access_data` (which will be 8 bytes long) and siblings;
* If it's a region read, we'll look at the target hash and siblings (so only look at the proof field);
*  If it's a write, we'll look at the target hash and siblings (so only look at the proof field).

Seems correct to you?

Cheers!

**Comment by diegonehab** [comment]
I  feel like a lot of  this discussion is based on taste, and we should focus on what is impossible to do in the current schema before we discuss taste. I don't agree on the difference between Ranges and non-Ranges.  I think a word is simply a special case of range where `log2_size == 3`. If you want to have a type where the read value is mandatory  rather than optional (dunno how to accomplish this without defactoring a bunch of fields), we can discuss that. But size doesn't matter. (He he he).

What is definitely missing right now is that writes assume the data will be there and, therefore, the hash can be computed from it. It turns out that we want the opposite: we want the hash itself to be mandatory but the value to be optional.

Unless we want to abandon the self-containment of the proofs or make them mandatory, there will be some repetition. You could increase repetition even more by adding the target hash from the proof outside it as well. Something like this

```
enum Access {
    proof: Option<Proof>,
    address: u64,
    log2_size: int,
    read_value: Option<Bytes>,
    read_hash: bytes32,
    Write {
        written_hash: bytes32,
        written_value: Option<Bytes>
    }
    Read {
    } 
}
```

Or 

enum Access {
    proof: Option<Proof>,
    address: u64,
    log2_size: int,
    Write {
        read_value: Option<Bytes>,
        read_hash: bytes32,
        written_hash: bytes32,
        written_value: Option<Bytes>
    }
    Read {
        value: Option<Bytes>,
        hash: bytes32,
    }
    ReadWithValue {
        value: Bytes,
        hash: bytes32,
    }  
}
```

**Comment by vfusco** [comment]
closed by #141
